File: src/main/java/CanvasBuilder.java
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressIndicatorProvider;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiMethod;
import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Set;

class CanvasBuilder {
    private ProgressIndicator progressIndicator;
    private Set<Dependency> dependencySnapshot = new HashSet<>();

    @NotNull
    Canvas build(@NotNull CanvasConfig canvasConfig) {
        // cancel existing progress if any
        if (this.progressIndicator != null) {
            this.progressIndicator.cancel();
        }
        this.progressIndicator = ProgressIndicatorProvider.getGlobalProgressIndicator();

        // build a dependency snapshot for the entire code base
        this.dependencySnapshot = Utils.getDependencySnapshot(canvasConfig, this.dependencySnapshot);

        // visualize the viewing part as graph
        Set<PsiFile> files = Utils.getSourceCodeFiles(canvasConfig);
        Set<PsiMethod> methods = Utils.getMethodsInScope(canvasConfig, files);
        Set<Dependency> dependencyView = Utils.getDependencyView(canvasConfig, methods, this.dependencySnapshot);
        Graph graph = buildGraph(methods, dependencyView);
        return renderGraphOnCanvas(canvasConfig.getCallGraphToolWindow(), graph);
    }

    @NotNull
    private Graph buildGraph(@NotNull Set<PsiMethod> methods, @NotNull Set<Dependency> dependencyView) {
        Graph graph = new Graph();
        methods.forEach(graph::addNode);
        dependencyView.forEach(dependency -> {
            graph.addNode(dependency.getCaller());
            graph.addNode(dependency.getCallee());
            graph.addEdge(dependency.getCaller(), dependency.getCallee());
        });
        Utils.layout(graph);
        return graph;
    }

    @NotNull
    private Canvas renderGraphOnCanvas(@NotNull CallGraphToolWindow callGraphToolWindow, @NotNull Graph graph) {
        Canvas canvas = new Canvas(callGraphToolWindow, graph);
        MouseEventHandler mouseEventHandler = new MouseEventHandler(canvas);
        canvas.addMouseListener(mouseEventHandler);
        canvas.addMouseMotionListener(mouseEventHandler);
        canvas.addMouseWheelListener(mouseEventHandler);
        return canvas;
    }
}

File: src/main/kotlin/callgraph/CanvasBuilder.kt
package callgraph

import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressIndicatorProvider
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiMethod

class CanvasBuilder {
    private var progressIndicator: ProgressIndicator? = null
    private var fileModifiedTimeCache = mapOf<PsiFile, Long>()
    private var dependenciesCache = emptySet<Dependency>()

    fun build(canvasConfig: CanvasConfig) {
        // cancel existing progress if any
        this.progressIndicator?.cancel()
        this.progressIndicator = ProgressIndicatorProvider.getGlobalProgressIndicator()

        // build a dependency snapshot for the entire code base
        val dependencies = getDependencies(canvasConfig, this.dependenciesCache, this.fileModifiedTimeCache)

        // visualize the viewing part as graph
        val sourceCodeRoots = Utils.getSourceCodeRoots(canvasConfig)
        val files = Utils.getSourceCodeFiles(canvasConfig.project, sourceCodeRoots)
        val methods = Utils.getMethodsInScope(canvasConfig, files)
        val dependencyView = Utils.getDependencyView(canvasConfig, methods, dependencies)
        val graph = buildGraph(methods, dependencyView)
        canvasConfig.canvas.reset(graph)
    }

    private fun buildGraph(methods: Set<PsiMethod>, dependencyView: Set<Dependency>): Graph {
        val graph = Graph()
        methods.forEach { graph.addNode(it) }
        dependencyView.forEach {
            graph.addNode(it.caller)
            graph.addNode(it.callee)
            graph.addEdge(it.caller, it.callee)
        }
        Utils.layout(graph)
        return graph
    }

    private fun getDependencies(
            canvasConfig: CanvasConfig,
            dependenciesCache: Set<Dependency>,
            fileModifiedTimeCache: Map<PsiFile, Long>
    ): Set<Dependency> {
        val allFiles = Utils.getAllSourceCodeFiles(canvasConfig.project)
        val newFiles = allFiles.filter { !fileModifiedTimeCache.containsKey(it) }
        val changedFiles = allFiles
                .filter { fileModifiedTimeCache.containsKey(it) && fileModifiedTimeCache[it] != it.modificationStamp }
                .toSet()
        val validDependencies = dependenciesCache
                .filter {
                    !changedFiles.contains(it.caller.containingFile) && !changedFiles.contains(it.callee.containingFile)
                }
                .toSet()
        val invalidFiles = dependenciesCache
                .filter { !validDependencies.contains(it) }
                .flatMap { listOf(it.caller.containingFile, it.callee.containingFile) }
                .toSet()
        val filesToParse = newFiles.union(invalidFiles)
        val methodsToParse = Utils.getMethodsFromFiles(filesToParse)

        // parse method dependencies
        canvasConfig.callGraphToolWindow.resetProgressBar(methodsToParse.size)
        val newDependencies = methodsToParse
                .flatMap {
                    canvasConfig.callGraphToolWindow.incrementProgressBar()
                    Utils.getDependenciesFromMethod(it)
                }
                .toSet()
        val dependencies = validDependencies.union(newDependencies)

        // cache the dependencies for next use
        this.dependenciesCache = dependencies
        this.fileModifiedTimeCache = allFiles.associateBy({ it }, { it.modificationStamp })

        return dependencies
    }
}
